### ISSUE TITLE
Classes not being added to groups

### ISSUE THREAD

**Issue by jwakely** [doxygen/doxygen#8638]
**Describe the bug**

C++ classes are sometimes not documented under the "module" that they are added to by using grouping commands such as `\defgroup` or `\addtogroup` or explicit `\ingroup` commands.

**Expected behavior**

In the attached reproducer, the `std::bad_exception` and `std::system_error` classes should be in the "Exceptions" module. The `std::error_code` and `std::error_condition` classes should be in the Diagnostics module. They are all documented (you can find them in the Class Index) but they are not grouped where they should be.

**To Reproduce**

[doxygen-groups-bug.tar.gz](https://github.com/doxygen/doxygen/files/6749766/doxygen-groups-bug.tar.gz)

Extract attached tarball, then run `doxygen`

**Version**

1.9.1

LSB Version: :core-4.1-amd64:core-4.1-noarch
Distributor ID: Fedora
Description: Fedora release 34 (Thirty Four)
Release: 34
Codename: ThirtyFour

Using the default Fedora package, doxygen-1.9.1-10.fc34.x86_64

**Comment by jwakely** [comment]
This is causing bugs in the docs for libstdc++ as noted at https://github.com/gcc-mirror/gcc/commit/6963c3b9ed1308bf5d2d8877956e56fec92853a7

Some classes that are defined in `<type_traits>` are correctly grouped, but a large number are not, and so don't appear in the relevant part of the docs. The only way to find them is the class index.

**Comment by jwakely** [comment]
It seems that adding `@class foo` to the comment block for `class foo` causes it to be correctly added to the group.

I don't know why that makes any difference, Doxygen is clearly able to tell that the comment block is for that class, because it puts the text on the page for the class. But if I tell it what it already knows, then this bug goes away.

**Comment by CambridgeDave** [comment]
Using the example from the documentation on grouping, I find that adding any non-whitespace content to a class stops it from being included in the group:

    /** @defgroup group1 The First Group
     *  This is the first group
     *  @{
     */
     
    /** @brief class C1 in group 1 - *is* documented in the group */
    class C1 {};
     
    /** @brief class C2 in group 1 - *is not* documented in the group */
    class C2 { /* This added */ };
     
    /** @} */ // end of group1

This behaviour started with v1.8.17.

**Comment by albert-github** [comment]
@CambridgeDave 

I tried the code you supplied with the 1.8.16, 1.8.17 and 1.9.1 versions of doxygen and the configuration file from the group example, but I don't see a difference. What did I oversee? (maybe enlighten with some images?)

Also (always): Can you please attach a, small, self contained example (source+configuration file in a tar or zip) that allows us to reproduce the problem? Please don't add external links as they might not be persistent.

**Comment by albert-github** [comment]
@jwakely 
I tried the example with the 1.9.1 version and with the current master (1.9.2 (97dffa0977cc41729942cf3a8c87232eb7edc09d)

For the Diagnostics and 1.9.1 I see (file html/a00023.html):
![image](https://user-images.githubusercontent.com/5223533/126293645-eed2136a-21d0-4905-a6a1-1b3a84fa9c68.png)

For the Diagnostics and master I see (file html/a00023.html):
![image](https://user-images.githubusercontent.com/5223533/126293726-be7587b5-2163-4184-96d6-c40e70a4ca61.png)

For the Exceptions and 1.9.1 I see (file html/a00026.html):
![image](https://user-images.githubusercontent.com/5223533/126293815-96aff627-00a3-427d-b2be-501568b2dcac.png)
(so no classes at all).

For the Exceptions and master I see (file html/a00026.html):
![image](https://user-images.githubusercontent.com/5223533/126293917-0dff0eaa-2762-410a-af07-5cb19671e89c.png)


Are these the problems you meant?

**Comment by jwakely** [comment]
@albert-github yes I can confirm the same results (I have a `master` build now, which I didn't have when (I reported this). So it seems to be fixed on master, that's good.

Is the root cause of the bug understood? Is the commit that fixed it known, so I can look into backporting it to the Fedora RPMs? Are there tests to prevent regressions?

**Comment by albert-github** [comment]
Setting the issue to fixed but not released.

- Is the root cause of the bug understood? Is the commit that fixed it known, so I can look into backporting it to the Fedora RPMs? 
  The first commit that the problem is fixed is (as far as I can see):
  ```
  Commit: b3cefe5954ed37ebcf843eced45e878b7b3dddcc [b3cefe5]
  Date: Thursday, February 4, 2021 9:11:15 PM

  issue #8371: @defgroup contained in the markdown mainpage.md are not including the @ingroup marked classes in the generated documentation.
  ```
  (looks like an unrelated problem to #8371 but it is not. Both issues used the `\defgroup` in combination with the ``ingroup`).
  My personal opinion: Regarding backporting, not really advised as at that moment it will be very hard to compare results in case of problems.
- Are there tests to prevent regressions?
  Not that I know of, feel free to create a test in the testing directory of doxygen.

**Comment by CambridgeDave** [comment]
> @albert-github 
> 
> I tried the code you supplied with the 1.8.16, 1.8.17 and 1.9.1 versions of doxygen and the configuration file from the group example, but I don't see a difference. What did I oversee? (maybe enlighten with some images?)

Thanks for looking at it.

I'm now worried mine is a separate issue, and I'm muddying the waters, but here is a tiny example that reproduces the problem.

[DoxygenGroupTest.tar.gz](https://github.com/doxygen/doxygen/files/6847840/DoxygenGroupTest.tar.gz)

Importantly, it looks like this only happens if `GROUP_NESTED_COMPOUNDS = YES` in the configuration.

**Comment by albert-github** [comment]
At first glance it does look like muddying the waters, though:

With 1.8.16 I see:
![image](https://user-images.githubusercontent.com/5223533/126317397-02c80236-b413-4625-b598-8329eeded1ff.png)

With 1.9.1 I see:
![image](https://user-images.githubusercontent.com/5223533/126317441-0ae0dec9-38ac-4806-887c-a3959dfcc517.png)


with the current master(1.9.2 (97dffa0977cc41729942cf3a8c87232eb7edc09d)) I see again:
![image](https://user-images.githubusercontent.com/5223533/126317577-6419ec5e-24e1-4e8f-be1a-db76c7f40e5e.png)

Also in this case the fix is by the
```
Commit: b3cefe5954ed37ebcf843eced45e878b7b3dddcc [b3cefe5]
Date: Thursday, February 4, 2021 9:11:15 PM

issue #8371: @defgroup contained in the markdown mainpage.md are not including the @ingroup marked classes in the generated documentation.
```
so there was no muddying, but it was helpful that it was with this issue...

**Comment by CambridgeDave** [comment]
Good news!  I'll build from source for now, then.  Cheers!

**Comment by jwakely** [comment]
@albert-github 

> My personal opinion: Regarding backporting, not really advised as at that moment it will be very hard to compare results in case of problems.

Understood. Thanks for the commit ID.

**Comment by doxygen** [comment]
This issue was previously marked 'fixed but not released',
which means it should be fixed in doxygen version 1.9.2.
Please verify if this is indeed the case. Reopen the
issue if you think it is not fixed and please include any additional information
that you think can be relevant (preferably in the form of a self-contained example).

**Comment by jwakely** [comment]
> The first commit that the problem is fixed is (as far as I can see):
>       ```
>       Commit: b3cefe5954ed37ebcf843eced45e878b7b3dddcc [b3cefe5]
>       Date: Thursday, February 4, 2021 9:11:15 PM
>       
>       issue #8371: @defgroup contained in the markdown mainpage.md are not including the @ingroup marked classes in the generated documentation.
>       ```
>       (looks like an unrelated problem to [@defgroup contained in the markdown mainpage.md are not including the @ingroup marked classes in the generated documentation. #8371](https://github.com/doxygen/doxygen/issues/8371) but it is not. Both issues used the `\defgroup` in combination with the `\ingroup`).

I bisected it and confirmed that b3cefe5954ed37ebcf843eced45e878b7b3dddcc fixed my issue.

I also determined that my issue only happened with `GROUP_NESTED_COMPOUNDS=YES` even though the classes that were missing from the group were not nested classes so should not have been affected by that option.
